Relfrozenxid must be invalid for append-optimized tables

asimrp opened the pull request:
Append-optimized tables do not contain transaction information in their tuples.  Therefore, `pg_class.relfrozenxid` must remain invalid. This is being done correctly during table creation, however, when the table was rewritten, the `relfrozenxid` was accidentally set.  Fix it now.

The fixme comments that led me to this bug can now be removed.

ashwinstar commented on src/test/regress/sql/create_am_gp.sql:
```diff
@@ -39,3 +39,15 @@ create table create_am_gp_ao6(a int, b int) using ao_row_testam with (compressty
 create table create_am_gp_ao7(a int, b int encoding (compresstype=zlib)) using ao_col_testam distributed by (a);
 \d+ create_am_gp_ao7
+
+-- Append optimized tables should not have a valid relfrozenxid in
```
Seems we have tests floating around in two tests already for relfrozenxid related on AO tables in `vacuum_gp.sql` and `freeze_aux_tables.sql`. One is testing vacuum specific activity and other freezing of auxiliary tables. Now we will have this third place. Will it be good to consolidate them in one place?

Also, do we need test for CO as well? or put in file where both will get tested by sql generation logic

ashwinstar reviewed (approved):
CI seems to be failing (maybe the newly added assert is hitting for some case). Otherwise look good to be :+1: apart from test comment.

asimrp commented:
Indeed, the PR pipeline indicates the newly added assertion is tripping.  And it highlighted a new bug: `alter table ao_row_table add column ... ` this rewrite is causing relfrozenxid to be set to a valid transaction ID.

There already is a fixme in `swap_relation_files()` that needs to be addressed in order to fix it:
```
        // GPDB_12_MERGE_FIXME                                                                                                                                                                                                                
        //Assert(should_have_valid_relfrozenxid(relform1->relkind));                                                                                                                                                                          
```

asimrp commented:
I've changed it so that the tests are moved to `uao_ddl/` to leverage AO row and column test generation, thank you Ashwin for the suggestion.  Also refactored the patch such that `should_have_valid_relfrozenxid()` function is no longer needed.  Please take another look.

ashwinstar commented on src/backend/utils/cache/relcache.c:
```diff
@@ -1812,6 +1812,12 @@ RelationInitTableAccessMethod(Relation relation)
+		 * Greenplum: append-optimized relations should not have a valid
+		 * relfrozenxid.
+		 */
+		Assert (!RelationIsAppendOptimized(relation) ||
```
Well if we really wish to live to the spirit of this comment (just for Assert we do or not depends :-))
```
/*                                                                                                                                          
 * CAUTION: this macro is a violation of the absraction that table AM and                                                                   
 * index AM interfaces provide.  Use of this macro is discouraged.  If                                                                      
 * table/index AM API falls short for your use case, consider enhancing the                                                                 
 * interface.                                                                                                                               
```
we should add a boolean variable to `TableAmRoutine` like `validRelfrozenxidExpected`. Set to true for heap and false for AO and CO table am handlers. That is then checked couple of lines below after calling `GetTableAmRoutine()` and completely avoiding usage of this macro

asimrp replied:
Let's track this usage of `RelationIsAppendOptimized()` macro in issue #11032 and leave the assert as is.  Is that good enough?

ashwinstar reviewed (approved):
Optimizer answer files also needs updating.

asimrp commented:
Thank you for checking the PR pipeline, optimizer answer files are updated.  I will push once the PR pieline is green.

asimrp commented:
PR pipeline has failed due to a pg_class <--> gpfastsequence foreign key check failure in gpcheckcat stage.  Hijacking the CI container did not yield much because of subsequent vacuum run as part of isolation2 suite of tests.
```
Found a total of 1 issue(s)

----------------------------------------------------
Relation oid: 49221
Relation schema: N/A
Relation name: N/A

    Name of test which found this issue: foreign_key_pg_class
        No pg_class {'oid': 49221} entry for gp_fastsequence {'objid': 49221, 'objmod': 2} on content 0 (b13863e2-2572-4ff0-6009-61b2cb737f51:70
```
Issue #11044 is created to track the assertion failure in gpcheckcat.  However, I don't see a way to deal with the foreign key test failure, besides keeping an eye on it.  It definitely is not related to the change proposed in this PR.

asimrp commented:
Merged to master as e68d5b8aeae9df366901e694666fe763015e9f60
